### PR TITLE
Serverless: add example with serverless

### PR DESCRIPTION
**Pull request by kamleshchandnani**
This PR addresses #6684

### CODE REVIEW

File: examples/with-serverless/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# nextjs-serverless
```
Comment (timneutkens) [inline examples/with-serverless/README.md]: I think the example should be called `with-serverless-framework`, there are many ways do to serverless (eg deploying to Now) and this community plugin is just one of them.

Comment (kamleshchandnani) [reply examples/with-serverless/README.md]: makes sense. Fixed ✅

File: examples/with-serverless/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# with-serverless-framework
```
Comment (timneutkens) [inline examples/with-serverless/README.md]: The directory should be updated too 💯

Comment (kamleshchandnani) [reply examples/with-serverless/README.md]: Done 💯

**Comment by lfades** [comment]
@kamleshchandnani We would prefer to have an example using [this plugin](https://github.com/danielcondemarin/serverless-nextjs-plugin), basically what it does is to create serverless functions for the pages and upload the static files to S3

**Comment by kamleshchandnani** [comment]
@timneutkens - I and @lfades were discussing this, So I need some custom things which cannot be done as part of that plugin and neither is documented anywhere else. I'll describe the motivation behind this example

I wanted to create a nextjs app which could achieve following things:
1. Work with `serverless` framework.
2. Able to use express for middlewares and hence this [file](https://github.com/zeit/next.js/pull/6685/files#diff-118d0a1bd8b974ed16f6041942978fd3). While this is just for now we have many middlewares some of them are custom as well. So basically to name a few cookieparser, cache, A/B, auth, logger and many more.
3. Co locating of routes and server stuff.
4. `serverless-offline` to emulate things before deploying.

And the main difference here is I need a thin orchestrating layer in between which can orchestrate routing along with middlewares capabilities.

So if you think of an express + nextjs app I guess this is how you'll achieve this. So I'm doing exactly this but building them as serverless and deploying it on AWS lambdas and API gateways.

**Comment by danielcondemarin** [comment]
> 2\. Able to use express for middlewares and hence this [file](https://github.com/zeit/next.js/pull/6685/files#diff-118d0a1bd8b974ed16f6041942978fd3). While this is just for now we have many middlewares some of them are custom as well. So basically to name a few `helmet`, `cookieparser`, `cache`, `A/B`, `auth`, `logger` and many more.

It's important to note first that next 8 serverless pages are self contained and ready to deploy on AWS Lambda, without having to use `express`.  

In a nutshell `express` gives you **routing** and **middlewares**.  The Routing you already get it from AWS API Gateway when you are declaring how to forward traffic to the Lambda, e.g.

**serverless.yml**
```yml
post:
    handler: post.handler
    events:
      - http:
          path: post/{slug}
          method: get
          request:
            parameters:
              paths:
                slug: true
```

Now the middleware part is more interesting. I've thought of adding support to the [plugin](https://github.com/danielcondemarin/serverless-nextjs-plugin) to let users hook up their own middlewares. However, I think that using the whole of `express` only to benefit from the middleware mechanism it's a bit overkill. Even if we were to use [connect](https://github.com/senchalabs/connect) directly, it has the concept of `routing` which you already get from the API Gateway.
A more appropiate middleware framework would be [middyjs](https://github.com/middyjs/middy). It's more lightweight and tailored for use in a Lambda environment.
I'll try and formalise some sort of proposal in the following days of how the plugin could support something like this, but would be great to hear what everyone else thinks first.

**Comment by romainquellec** [comment]
I guess both is great as long as you dont have a huge project.
With this approach, you can hit the lambda size limit (75mo) pretty quick.
@danielcondemarin, you should also PR with your plugin as you know what you re doing with it 👍

**Comment by kamleshchandnani** [comment]
> With this approach, you can hit the lambda size limit (75mo) pretty quick.

@romainquellec AFAIK for AWS the deployment package size is 50MB compressed and 250MB if you upload it to S3 and if you use `serverless` it always uploads things onto S3. So we are good with that 🙂

**Comment by kamleshchandnani** [comment]
> @kamleshchandnani We would prefer to have an example using [this plugin](https://github.com/danielcondemarin/serverless-nextjs-plugin), basically what it does is to create serverless functions for the pages and upload the static files to S3

So the design that this plugin fulfils and the the solution that this PR addresses serves different purpose IMHO as mentioned [here](https://github.com/zeit/next.js/pull/6685#issuecomment-473747387)

**Comment by romainquellec** [comment]
> AFAIK for AWS the deployment package size is 50MB compressed and 250MB if you upload it to S3 and if you use serverless it always uploads things onto S3. So we are good with that 

I'm using your approach, and with serverless framework, i'm sometimes hitting the limit with >75mo.
(Officially, Limit is 50mo, but you can do more). For my case, Serverless upload to S3, but compressed, so I'm hitting the first limit.

**Comment by mitchellhuang** [comment]
I actually like this example better than [serverless-nextjs-plugin](https://github.com/danielcondemarin/serverless-nextjs-plugin) because all pages are bundled into a single function. This reduces the number of cold starts on less trafficked pages since everything is loaded in a hot function. Obviously this depends on your traffic volume and how you want your application to scale on Lambda (all pages vs per page).

**Comment by timneutkens** [comment]
I'm going to close this PR as it does not illustrate best practices for the serverless target.

> I actually like this example better than serverless-nextjs-plugin because all pages are bundled into a single function.

Bundling into a single function is bad in pretty much all cases. Cold boot is not noticeable when lambda size is <5Mb. You'll always have cold starts. And even hot functions get stopped etc.

Furthermore using `express` for the functions will slow them down / increase bundle size.

Also bundling everything into one function means that say you have 10 pages, you'll have 10 versions of React into the bundle. Same for every component etc.

**Pull request closed without merging**